### PR TITLE
feat(terraform): Added base class for cloudsplaining iam checks to be integrated between data and resource objects

### PR DESCRIPTION
**Pull request by bo156**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    #    
    Allowed prefixs:
    ansible|argo|arm|azure|bicep|bitbucket|circleci|cloudformation|dockerfile|github|gha|gitlab|helm|kubernetes|kustomize|openapi|sca|secrets|serverless|terraform|general|graph|terraform_plan
    #
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

Added a base class for handling `clousplaining` checks, to allow future support of both `resource` and `data` checks without code duplication.

Fixes # (issue)

### Description
*Include a description of what makes it a violation and any relevant external links.*

### Fix
*How does someone fix the issue in code and/or in runtime?*

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my feature, policy, or fix is effective and works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by tronxd — APPROVED** [review]

**Review by achiar99 — APPROVED** [review]

File: checkov/terraform/checks/data/base_cloudsplaining_data_iam_check.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+from checkov.common.models.enums import CheckResult, CheckCategories
+from checkov.terraform.checks.data.base_check import BaseDataCheck
+from checkov.terraform.checks.utils.base_cloudsplaining_iam_scanner import BaseTerraformCloudsplainingIAMScanner
+from checkov.terraform.checks.utils.iam_terraform_document_to_policy_converter import (
+    convert_terraform_conf_to_iam_policy,
+)
+
+
+class BaseTerraformCloudsplainingDataIAMCheck(BaseDataCheck, BaseTerraformCloudsplainingIAMScanner, ABC):
```
Comment (gruebel) [inline checkov/terraform/checks/data/base_cloudsplaining_data_iam_check.py:14]: not sure, why you subclass `ABC` here? there are no abstract annotations used.

File: checkov/terraform/checks/resource/base_cloudsplaining_resource_iam_check.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+from typing import Dict, List, Any
+
+from cloudsplaining.scan.policy_document import PolicyDocument
+
+from checkov.common.models.enums import CheckResult, CheckCategories
+from checkov.terraform.checks.resource.base_resource_check import BaseResourceCheck
+from checkov.terraform.checks.utils.base_cloudsplaining_iam_scanner import BaseTerraformCloudsplainingIAMScanner
+
+
+class BaseTerraformCloudsplaininResourceIAMCheck(BaseResourceCheck, BaseTerraformCloudsplainingIAMScanner, ABC):
```
Comment (gruebel) [inline checkov/terraform/checks/resource/base_cloudsplaining_resource_iam_check.py:11]: same here

File: checkov/terraform/checks/utils/base_cloudsplaining_iam_scanner.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import json
+import logging
+from abc import abstractmethod
+from typing import Dict, List, Any, Union
+
+from checkov.common.models.enums import CheckResult
+
+from cloudsplaining.scan.policy_document import PolicyDocument  # noqa: TC002
```
Comment (gruebel) [inline checkov/terraform/checks/utils/base_cloudsplaining_iam_scanner.py:8]: you can actually move it under `TYPE_CHECKING`, but then you also have to import `annotations`

**Review by gruebel — COMMENTED** [review]
in general, good job 💪

File: checkov/terraform/checks/utils/base_cloudsplaining_iam_scanner.py (hunk 3 above)
Comment (bo156) [reply checkov/terraform/checks/utils/base_cloudsplaining_iam_scanner.py:8]: cool, fixing now

File: checkov/terraform/checks/data/base_cloudsplaining_data_iam_check.py (hunk 1 above)
Comment (bo156) [reply checkov/terraform/checks/data/base_cloudsplaining_data_iam_check.py:14]: Actually we also need to implement `cloudformation_analysis` which is only in the check itself

File: checkov/terraform/checks/resource/base_cloudsplaining_resource_iam_check.py (hunk 2 above)
Comment (bo156) [reply checkov/terraform/checks/resource/base_cloudsplaining_resource_iam_check.py:11]: same

**Pull request merged**